README.md
# Facebook Ads Performance Downloader

A Python script for downloading structure and performance data from ad accounts in a single [business manager](https://developers.facebook.com/docs/apps/business-manager) using the Facebook Ads API ([v3.2](https://developers.facebook.com/docs/marketing-api/reference/v3.2)) to local files. 

The [mara Facebook ads performance pipeline](https://github.com/mara/facebook-ads-performance-pipeline) can be, then, used for loading and transforming the downloaded data into a dimensional schema.

## Resulting data
By default, it creates two data sets:

1. **Ad Performance** consists of measures such as impressions, clicks, conversions and spend. The script creates one sqlite3 database per day and ad account in a specified time range:

        data/2017/01/01/facebook/ad-performance-act-1234567890-v2.sqlite3
        data/2017/01/02/facebook/ad-performance-act-1234567890-v2.sqlite3

    For the last 28 days, the script always redownloads the files as data still changes (e.g spend or attributed conversions). Beyond that, files are only downloaded for dates where the ad account existed but the file is missing. 
    **Note**: If you are using an attribution window larger than 28 days adjust the `redownload_window` config accordingly.
    
    The resulting sqlite3 databases contain a single table `ad_performance`, with individual rows per ad and the device and placement breakdown. See [Insights Breakdowns](https://developers.facebook.com/docs/marketing-api/insights/breakdowns/v3.2) for a documentation of the breakdown. The performance data is stored in JSON format in a single column

        date        | 2017-01-01
        ad_id       | 1234567890
        device      | desktop
        placement   | desktop_feed
        performance | {"spend": 0.42,  
                    |  "impressions": 25,
                    |  "actions": [{"value": 1.0, "28d_click": 1.0, "action_type": "link_click"}, 
                    |              {"value": 20.0, "28d_click": 20.0, "action_type": "offsite_conversion.fb_pixel_add_to_cart"}, 
                    |              {"value": 2.0, "28d_click": 2.0, "action_type": "offsite_conversion.fb_pixel_purchase"}, 
                    |              {"value": 18.0, "28d_click": 17.0, "action_type": "offsite_conversion.fb_pixel_view_content"}, 
                    |              {"value": 1.0, "28d_click": 1.0, "action_type": "page_engagement"}, 
                    |              {"value": 1.0, "28d_click": 1.0, "action_type": "post_engagement"}, 
                    |              {"value": 45.0, "28d_click": 34.0, "action_type": "offsite_conversion"}], 
                    |  "action_values": [{"value": 2.0, "28d_click": 2.0, "action_type": "offsite_conversion.fb_pixel_purchase"}, 
                    |                    {"value": 2.0, "28d_click": 2.0, "action_type": "offsite_conversion"}]}
                    
    See [Insights Fields](https://developers.facebook.com/docs/marketing-api/insights/fields) for a documentation of the fields.
    
2. **Account Structure** information. This sqlite3 database containes one table `account_structure` which is always upserted by the script:

        data/facebook-account-structure-v2.sqlite3

    Each rows contains one ad together with its ad group, campaign and account:
    
        ad_id         | 1234567890
        ad_name       | Foo
        ad_group_id   | 3456789012
        ad_group_name | Bar
        campaign_id   | 5678901234
        campaign_name | Baz
        account_id    | 7890123456
        account_name  | Qux
        attributes    | {"Target": "buyer", "Channel": "reactivation", "Country": "France", "Language": "French"}

    The `attributes` field contains a JSON representation of all [labels](https://support.google.com/adwords/answer/2475865) of an ad or its parents in a `{Key=Value}` syntax. For example an account has the label `{Channel=reactivation}`, then all ads below will have the `"Channel": "reactivation"`. 
    **Note**: Labels on lower levels overwrite those from higher levels.


## Getting Started

### Prerequisites

To use the Facebook Ads Performance Downloader you have to

- consolidate all accounts of a company under a single [business manager](https://developers.facebook.com/docs/apps/business-manager), 
- create a normal [system user](https://developers.facebook.com/docs/marketing-api/businessmanager/systemuser) for that business manager using the guide in the link or the [System User page](https://business.facebook.com/settings/system-users) in your business manager.
- give the system user access to all ad accounts that should be downloaded,
- create an [app](https://developers.facebook.com/docs/apps/register#step-by-step-guide) and disable development mode,
- set up app and system user credentials to access the Facebook Ads API. See [Set up your credentials](#set-up-your-credentials) for the necessary steps.

Optionally, you can apply labels on all hierarchy levels for segmenting the structure.

### Installation

 The Facebook Ads Performance Downloader requires:

    Python (>= 3.5)
    facebook_business (==3.0.0)
    click (>=6.0)

The easiest way to install facebook-ads-downloader is using pip

    pip install git+https://github.com/mara/facebook-ads-performance-downloader.git

In case you want to install it in a virtual environment:

    $ git clone git@github.com:mara/facebook-ads-performance-downloader.git facebook_downloader
    $ cd facebook_downloader
    $ python3 -m venv .venv
    $ .venv/bin/pip install .

### Set up your credentials

Find the `app_id` and `app_secret` in your app's dashboard at `https://developers.facebook.com/apps/<YOUR_APP_ID>/dashboard/`:
![Facebook App Dashboard](docs/facebook-app-dashboard.png)

Log into your Facebook account and go to the system users page of your company's business manager at `https://business.facebook.com/settings/system-users` and click on *Generate New Token*:
![System Users](docs/facebook-business-manager-system-users.png)

Select the app you want to use to download the data, activate the `read_insights` and `ads_read` scope and generate the token:
![Facebook System User Token Generation](docs/facebook-system-user-token-scope.png)

The generated `access_token` is then shown to you:
![Facebook System User Access Token](docs/facebook-system-user-access-token.png)

Currently Facebook requires you to perform a number of successful API calls before being able to access more than 5 accounts. If that's the case use the `target_accounts` parameter to specify these 5 accounts and avoid API errors.

### Refresh API Token
Facebook invalidates the API token once every couple of months. To refresh it, go to `https://developers.facebook.com/tools/explorer/<YOUR_APP_ID>`:
![Facebook API Token Refresh](docs/facebook-system-user-token-refresh.png)
Click on `Get User Access Token`, select `read_insights` and `ads_read`, and generate the new token, which will appear in the line next to the button pressed on the screenshot. This will generate a short-lived access token. To extend it, go to `https://developers.facebook.com/tools/debug/accesstoken/?access_token=<GENERATED_SHORT_LIVED_TOKEN>` (or click on the (i)nformation icon next to the token and choose _Open in Access Token Tool_) and click on _Extend Access Token_ button. The new long-lived access token will be displayed below.

## Usage

To run the Facebook Ads Performance Downloader call `download-facebook-performance-data` with its config parameters:  

    $ download-facebook-performance-data --app_id 1234567890 \
    --app_secret ABCDEFEGHIJKL \
    --access_token aBcDeFgHiJkLmNoPqRsTuVwXyZ \
    --data_dir /tmp/facebook

For all options, see 

    $ download-facebook-performance-data --help
    Usage: download-facebook-performance-data [OPTIONS]
    
      Downloads data. When options are not specified, then the defaults from
      config.py are used.
    
    Options:
      --app_id TEXT             The app id obtained from the app's settings in
                                facebook for developers
                                
                                https://developers.facebo
                                ok.com/apps/<APP_ID>/settings/. Example:
                                "1234567890"
      --app_secret TEXT         The app secret obtained from the app's settings in
                                facebook for developers
                                
                                https://developers.facebo
                                ok.com/apps/<APP_ID>/settings/. Example: "aBcDeFg"
      --access_token TEXT       The access token of the system user with the
                                following credentials:
                                - read_insights
                                - ads_read
                                https://business.facebook.com/settings/system-
                                users/<SYSTEM_USER_ID>?business_id=<BUSINESS_ID>.
                                Example: "foo"
      --data_dir TEXT           The directory where result data is written to.
                                Example: "/tmp/facebook_ads"
      --first_date TEXT         The first day for which data is downloaded.
                                Example: "2015-01-01"
      --redownload_window TEXT  The number of days for which the performance data
                                will be redownloaded. Example: "28"
      --target_accounts TEXT    The accounts to download, comma separated, if not
                                set or empty each available account will be tried.
                                Example: ""
      --number_of_ad_performance_threads TEXT
                                The number of threads used to download ad
                                performance. Example: "10"
      --help                    Show this message and exit.


CHANGELOG.md
# Changelog

## 3.3.0 (2021-10-11)

- Switch to current Facebook API version (11.0.0)

## 3.2.0 (2021-04-19)

- Switch to current Facebook API version (10.0.0)

## 3.1.0 (2020-04-06)

- Switch to current Facebook API version (6.0.0) 


## 3.0.1 (2019-08-20)
- Switch to current Facebook API version (4.0.2) 
- Replaced deprecated function remote_read with api_get

## 3.0.0 (2019-04-13)

- Change MARA_XXX variables to functions to delay importing of imports

**required changes** 

- If used together with a mara project, Update `mara-app` to `>=2.0.0`


## 2.1.0 (2019-02-06)
- Switch to current Facebook API version (3.2.4)
  - Fix python3.7 compatibility (usage of `async` as variable name)
- Block (sleep) worker threads when request limit reached
- Fix potential deadlocks in the early exit case
  - Logging for failures now also prints the full stack trace.

## 2.0.0 - 2.0.1 (2018-09-27)

- Switch to current Facebook API version (3.0.0)
- Parallelize ad performance download
- Unify output file names, bump version to 'v2'
- Fix API bug "2601" on handling async requests

**required changes**

- Adapt ETL to new output file names


## 1.5.0 - 1.5.3 (2018-07-24)

- Fix facebook-api request and rate limit issues with asynchronous requests
- Add check for 'Job Failed' status
- Describe how to refresh the API token
- Make action_attribution_windows parameter configurable

## 1.4.0 - 1.4.3 (2017-12-28)

- Updated `facebookads` to 2.11.1`
- catch not existing impressions and spend
- small bug fix for logging with target accounts 
- Insert unknown device if device is empty
- allow more characters in labels
- Handle 'unknown error' by retrying on all FacebookRequestError exceptions

## 1.3.0 (2017-10-05)

- Add the `target_accounts` parameter to optionally limit the accounts to download
- Updated `facebookads` to 2.10.1`

## 1.2.0 (2017-09-21)

- Made the config and click commands discoverable in [mara-app](https://github.com/mara/mara-app) >= 1.2.0


## 1.1.0 - 1.1.2 (2017-05-17)

- Updated to version 2.9.1
- Drop deprecated Placement dimension 
- Updated to version 2.9.2
- Download performance data for ads in all states except deleted ones

**required changes**

- Remove work files and run import again after downloading latest files 



## 1.0.0 - 1.0.2 (2017-03-01) 

- Initial version
- made cli and config discoverable
- Changed logic for first date. Resolves problems for the case of absent creation time for an account.


facebook_downloader/cli.py
"""Command line interface for facebook downloader"""

import logging
from functools import partial

import click
from facebook_downloader import config


def config_option(config_function):
    """Helper decorator that turns an option function into a cli option"""
    return (lambda function: click.option('--' + config_function.__name__,
                                          help=config_function.__doc__.strip() + '. Example: "' +
                                               str(config_function()) + '"')(function))


def apply_options(kwargs):
    """Applies passed cli parameters to config.py"""
    for key, value in kwargs.items():
        if value: setattr(config, key, partial(lambda v: v, value))


@click.command()
@config_option(config.app_id)
@config_option(config.app_secret)
@config_option(config.access_token)
@config_option(config.data_dir)
@config_option(config.first_date)
@config_option(config.redownload_window)
@config_option(config.target_accounts)
@config_option(config.number_of_ad_performance_threads)
def download_data(**kwargs):
    """
    Downloads data.
    When options are not specified, then the defaults from config.py are used.
    """
    from facebook_downloader import downloader

    apply_options(kwargs)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
    downloader.download_data()


facebook_downloader/config.py
"""
Configures access to the Facebook Ads API and where to store results
"""


def data_dir() -> str:
    """The directory where result data is written to"""
    return '/tmp/facebook_ads'


def first_date() -> str:
    """The first day for which data is downloaded"""
    return '2015-01-01'


def app_id() -> str:
    """The app id obtained from the app's settings in facebook for developers

    https://developers.facebook.com/apps/<APP_ID>/settings/
    """
    return '1234567890'


def app_secret() -> str:
    """The app secret obtained from the app's settings in facebook for developers

    https://developers.facebook.com/apps/<APP_ID>/settings/
    """
    return 'aBcDeFg'


def access_token() -> str:
    """The access token of the system user with the following credentials:
    - read_insights
    - ads_read

    https://business.facebook.com/settings/system-users/<SYSTEM_USER_ID>?business_id=<BUSINESS_ID>
    """
    return 'foo'


def redownload_window() -> str:
    """The number of days for which the performance data will be redownloaded"""
    return '28'


def target_accounts() -> str:
    """The accounts to download, comma separated, if empty each available account will be tried"""
    return ''


def action_attribution_windows() -> str:
    """The action attribution windows parameter. Default as '28d_click'
    https://developers.facebook.com/docs/marketing-api/insights/#sample"""
    return ['28d_click']
